### PR TITLE
Improve kubectl error messages

### LINKED ISSUE #31267: Improve kubectl error messages
List here:
https://github.com/kubernetes/community/wiki/Roadmap:-kubectl

Cross them out as you complete them

---
## List of additional issues
- [ ] #34457  exec pods on master unsupported
- [ ] #30081 --dry-run + -i unsupported

### LINKED ISSUE #34457: `kubectl exec` for pods running on the master return obscure error message
When try to use `kubectl exec` for the pods on the master node, got below error message:

```
error: unable to upgrade connection: 404 page not found
```

After did a bit searching, found this related issue: #29480, saying that on the master kubelet is running with `--enable-debugging-handlers=false`

If `kubectl exec` is not allowed on master, could we make the error message clear?

### PR DESCRIPTION
**Pull request by lichen2013**
Fixes #34457
Part of work on #31267

### CODE REVIEW

**Comment by MHBauer** [comment]
/ok-to-test

I'm no expert here, but it looks sane enough when compared with the existing `InstallDebuggingHandlers` part.

File: pkg/kubelet/server/server.go [hunk 1]
```diff
@@ -706,6 +779,11 @@ func (s *Server) getRun(request *restful.Request, response *restful.Response) {
 	writeJsonResponse(response, data)
 }
 
+func (s *Server) debuggingDisabled(request *restful.Request, response *restful.Response) {
+	message := []byte("Server endpoints for log collection and local running of containers and commands are disabled.")
+	writeJsonResponse(response, message)
```
Comment (CaoShuFeng) [inline pkg/kubelet/server/server.go]: What's the response status code after this change?
404 or something?

Comment (CaoShuFeng) [reply pkg/kubelet/server/server.go]: Seems like it will be 200 (statusOK).

Comment (CaoShuFeng) [reply pkg/kubelet/server/server.go]: And I think we should pass a right `json` to function write`Json`Response

File: pkg/kubelet/server/server.go [hunk 2]
```diff
@@ -417,6 +419,77 @@ func (s *Server) InstallDebuggingHandlers(criHandler http.Handler) {
 	}
 }
 
+// InstallDebuggingDisabledHandlers registers the HTTP request patterns that provider better error message
+func (s *Server) InstallDebuggingDisabledHandlers() {
+	glog.Infof("Adding debug disabled handlers to kubelet server.")
+
+	ws := new(restful.WebService)
```
Comment (CaoShuFeng) [inline pkg/kubelet/server/server.go]: I am not familiar with this part. So this is just a weak suggestion:
I think we can  simply route all subrequets of `/run`, `/exec`... to debugingDisabled

**Review by CaoShuFeng — CHANGES_REQUESTED** [review]

File: pkg/kubelet/server/server.go (hunk 1 above)
Comment (CaoShuFeng) [reply pkg/kubelet/server/server.go]: Maybe:
```
func (s *Server) debuggingDisabled(response *restful.Response) {
	message := "Server endpoints for log collection and local running of containers and commands are disabled."
	response.WriteError(http.StatusMethodNotAllowed, fmt.Errorf(message))   // not sure 405 is ok here, or 404?
}

```

Comment (lichen2013) [reply pkg/kubelet/server/server.go]: I think 405 is the right code here.

**Comment by CaoShuFeng** [comment]
Please squash three commits into one single commit.

File: pkg/kubelet/server/server.go [hunk 3]
```diff
@@ -417,6 +419,20 @@ func (s *Server) InstallDebuggingHandlers(criHandler http.Handler) {
 	}
 }
 
+// InstallDebuggingDisabledHandlers registers the HTTP request patterns that provider better error message
```
Comment (CaoShuFeng) [inline pkg/kubelet/server/server.go]: nit: provider-> provide

Comment (lichen2013) [reply pkg/kubelet/server/server.go]: Done!

**Comment by CaoShuFeng** [comment]
Looks good to me and test OK in my environment.

/assign @Random-Liu 
For approve.

File: pkg/kubelet/server/server.go [hunk 4]
```diff
@@ -417,6 +419,20 @@ func (s *Server) InstallDebuggingHandlers(criHandler http.Handler) {
...
 
+// InstallDebuggingDisabledHandlers registers the HTTP request patterns that provide better error message
+func (s *Server) InstallDebuggingDisabledHandlers() {
+	glog.Infof("Adding debug disabled handlers to kubelet server.")
+
+	h := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		http.Error(w, "Server endpoints for log collection and local running of containers and commands are disabled.", http.StatusMethodNotAllowed)
+	})
+
+	paths := []string{"/run/", "/exec/", "/attach/", "/portForward/", "/containerLogs/", "/runningpods/"}
```
Comment (tallclair) [inline pkg/kubelet/server/server.go]: This might as well include all the debug paths. You're missing: pprofBasePath, logsPath

File: pkg/kubelet/server/server.go [hunk 5]
```diff
@@ -417,6 +419,20 @@ func (s *Server) InstallDebuggingHandlers(criHandler http.Handler) {
...
+	for _, p := range paths {
+		s.restfulCont.Handle(p, h)
```
Comment (tallclair) [inline pkg/kubelet/server/server.go]: Does this also handle sub paths? E.g. `/run/foo`?

File: pkg/kubelet/server/server.go [hunk 6]
```diff
@@ -417,6 +419,20 @@ func (s *Server) InstallDebuggingHandlers(criHandler http.Handler) {
 	}
 }
 
+// InstallDebuggingDisabledHandlers registers the HTTP request patterns that provide better error message
+func (s *Server) InstallDebuggingDisabledHandlers() {
+	glog.Infof("Adding debug disabled handlers to kubelet server.")
```
Comment (tallclair) [inline pkg/kubelet/server/server.go]: Nit: drop this log message, it's sort of confusing

File: pkg/kubelet/server/server.go [hunk 7]
```diff
@@ -417,6 +419,20 @@ func (s *Server) InstallDebuggingHandlers(criHandler http.Handler) {
...
+
+	h := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		http.Error(w, "Server endpoints for log collection and local running of containers and commands are disabled.", http.StatusMethodNotAllowed)
```
Comment (tallclair) [inline pkg/kubelet/server/server.go]: I would reword this to: `Debug endpoints are disabled.`

**Review by tallclair — COMMENTED** [review]
Thanks. Please add some unit tests.

File: pkg/kubelet/server/server.go (hunk 6 above)
Comment (lichen2013) [reply pkg/kubelet/server/server.go]: Done!

File: pkg/kubelet/server/server.go (hunk 7 above)
Comment (lichen2013) [reply pkg/kubelet/server/server.go]: Done.

File: pkg/kubelet/server/server.go (hunk 4 above)
Comment (lichen2013) [reply pkg/kubelet/server/server.go]: Done.

File: pkg/kubelet/server/server.go (hunk 5 above)
Comment (lichen2013) [reply pkg/kubelet/server/server.go]: Yes, we add path by a string like "/run/" with a "/" at the end. It should be able to take care of sub paths.

And, I tested with "kubectl  exec" commands, it actually doing post to path "/exec/${namespace}/${pod}/${container}?command=${cmd}...", it works as expected.

File: pkg/kubelet/server/server.go [hunk 8]
```diff
@@ -417,6 +419,21 @@ func (s *Server) InstallDebuggingHandlers(criHandler http.Handler) {
...
+// InstallDebuggingDisabledHandlers registers the HTTP request patterns that provide better error message
+func (s *Server) InstallDebuggingDisabledHandlers() {
+
+	h := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		http.Error(w, "Debug endpoints are disabled.", http.StatusMethodNotAllowed)
+	})
+
+	paths := []string{
+		"/run/", "/exec/", "/attach/", "/portForward/", "/containerLogs/",
+		"/runningpods/", "/pprofBasePath/", "/logsPath/"}
```
Comment (tallclair) [inline pkg/kubelet/server/server.go]: logsPath & pprofBasePath are variables. It should be:
```go
"/runningpods/", pprofBasePath, logsPath}
```

File: pkg/kubelet/server/server.go [hunk 9]
```diff
@@ -417,6 +419,21 @@ func (s *Server) InstallDebuggingHandlers(criHandler http.Handler) {
 	}
 }
 
+// InstallDebuggingDisabledHandlers registers the HTTP request patterns that provide better error message
+func (s *Server) InstallDebuggingDisabledHandlers() {
+
```
Comment (tallclair) [inline pkg/kubelet/server/server.go]: nit: extra blank line

File: pkg/kubelet/server/server_test.go [hunk 10]
```diff
@@ -204,7 +204,13 @@ type serverTestFramework struct {
 	criHandler      *utiltesting.FakeHandler
 }
 
-func newServerTest() *serverTestFramework {
+func newServerTest(disableDebug ...bool) *serverTestFramework {
+
```
Comment (tallclair) [inline pkg/kubelet/server/server_test.go]: nit: delete this blank line

File: pkg/kubelet/server/server_test.go [hunk 11]
```diff
@@ -204,7 +204,13 @@ type serverTestFramework struct {
 	criHandler      *utiltesting.FakeHandler
 }
 
-func newServerTest() *serverTestFramework {
+func newServerTest(disableDebug ...bool) *serverTestFramework {
```
Comment (tallclair) [inline pkg/kubelet/server/server_test.go]: Why is `disableDebug` a slice?

File: pkg/kubelet/server/server_test.go [hunk 12]
```diff
@@ -1635,3 +1641,30 @@ func TestCRIHandler(t *testing.T) {
...
 }
+
+func TestDebuggingDisabledHandlers(t *testing.T) {
+	fw := newServerTest(false)
+	defer fw.testHTTPServer.Close()
+
+	paths := []string{
+		"/run", "/exec", "/attach", "/portForward", "/containerLogs", "/runningpods",
+		"/run/", "/exec/", "/attach/", "/portForward/", "/containerLogs/", "/runningpods/",
+		"/run/xxx", "/exec/xxx", "/attach/xxx",
```
Comment (tallclair) [inline pkg/kubelet/server/server_test.go]: add `"/debug/pprof/profile"` and `"/logs/kubelet.log"`

File: pkg/kubelet/server/server.go (hunk 9 above)
Comment (lichen2013) [reply pkg/kubelet/server/server.go]: done.

File: pkg/kubelet/server/server.go (hunk 8 above)
Comment (lichen2013) [reply pkg/kubelet/server/server.go]: o... really sorry for this.

File: pkg/kubelet/server/server_test.go (hunk 11 above)
Comment (lichen2013) [reply pkg/kubelet/server/server_test.go]: What I want to do is adding disableDebug with default value 'true' to the func newServerTest, then there would be no need to change other places. 
But golang do not support default value.
I don't want to change other places from newServerTest() to newServerTest(true), and the slice make the inputs "optional".

File: pkg/kubelet/server/server_test.go (hunk 12 above)
Comment (lichen2013) [reply pkg/kubelet/server/server_test.go]: Done.

File: pkg/kubelet/server/server_test.go [hunk 13]
```diff
@@ -1635,3 +1640,30 @@ func TestCRIHandler(t *testing.T) {
...
+		require.NoError(t, err)
+		assert.Equal(t, "Debug endpoints are disabled.\n", string(body))
+
+		resp, err = http.Post(fw.testHTTPServer.URL+p, "", nil)
+		require.NoError(t, err)
+		assert.Equal(t, http.StatusMethodNotAllowed, resp.StatusCode)
+		body, err = ioutil.ReadAll(resp.Body)
+		require.NoError(t, err)
+		assert.Equal(t, "Debug endpoints are disabled.\n", string(body))
+	}
```
Comment (CaoShuFeng) [inline pkg/kubelet/server/server_test.go]: Would be better to test that default handlers are still accessable, when debug handlers are disabled.

File: pkg/kubelet/server/server_test.go (hunk 11 above)
Comment (CaoShuFeng) [reply pkg/kubelet/server/server_test.go]: Changeing newServerTest() to newServerTest(true) would do much harm, I think.

Comment (lichen2013) [reply pkg/kubelet/server/server_test.go]: There're 27 places, and possible most of these places do not actually care about the flag value.

Comment (CaoShuFeng) [reply pkg/kubelet/server/server_test.go]: Haha, OK.
Let me ask another question:
Why disableDebug is a slice of bool? Why not a slice of struct{}

Comment (lichen2013) [reply pkg/kubelet/server/server_test.go]: because I still want it indicate a "flag". slice is just a way to make "optional".

To be fair, I do not really understand why golang do not support default value.

File: pkg/kubelet/server/server_test.go (hunk 13 above)
Comment (lichen2013) [reply pkg/kubelet/server/server_test.go]: o, maybe update existing test cases with => fw := newServerTest(false) ???

Ha, this gives a good reason to use fw := newServerTest(true).

Comment (CaoShuFeng) [reply pkg/kubelet/server/server_test.go]: Cant understand it.

File: pkg/kubelet/server/server_test.go (hunk 11 above)
Comment (tallclair) [reply pkg/kubelet/server/server_test.go]: The way this is typically done in go is to add a new method with the parameter, and delegate to it from the old, i.e.
```go
func newServerTestWithDebug(enableDebugging bool) *serverTestFramework {
...
}

func newServerTest() *serverTestFramework {
  return newServerTestWithDebug(true)
}
```

File: pkg/kubelet/server/server_test.go (hunk 13 above)
Comment (tallclair) [reply pkg/kubelet/server/server_test.go]: I think it would be sufficient to just call a couple of the other paths, and verify that you get a 200 response code.

File: pkg/kubelet/server/server_test.go (hunk 11 above)
Comment (lichen2013) [reply pkg/kubelet/server/server_test.go]: o! Thanks a lot for this! Let me update the PR.

File: pkg/kubelet/server/server_test.go (hunk 13 above)
Comment (lichen2013) [reply pkg/kubelet/server/server_test.go]: Done!
